File: src/DiagToolsValidator/DiagToolsValidator.Core/Configuration/DiagToolsTest/DiagToolsTestRunConfiguration.fs
namespace DiagToolsValidator.Core.Configuration

open System
open System.IO
open System.Text
open System.Collections
open System.Collections.Generic
open System.Runtime.InteropServices

open YamlDotNet.Serialization

open DiagToolsValidator.Core.CoreFunctionality


module DiagToolsTestConfiguration =
    type SystemInformation() =
        member val OSName: string = null with get, set
        member val CPUArchitecture: string = null with get, set
        member val EnvironmentVariables = new Dictionary<string, string>() with get, set
        member val CLIDebugger: string = null with get, set
        
    type DotNetSetting() =
        member val SDKVersion: string = null with get, set
        member val DotNetRoot: string = null with get, set

    type DiagToolSetting() =
        member val DiagToolVersion: string = null with get, set
        member val ToolsToTest: string list = List.empty with get, set 
        member val ToolRoot: string = null with get, set
        member val Feed: string = null with get, set

    type TargetAppSetting() =
        member val BuildConfig: string = null with get, set
        member val ConsoleApp: DotNetApp.DotNetApp = new DotNetApp.DotNetApp() with get, set
        member val WebApp: DotNetApp.DotNetApp = new DotNetApp.DotNetApp() with get, set
        member val GCDumpPlayground: DotNetApp.DotNetApp = new DotNetApp.DotNetApp() with get, set

    type DiagToolsTestConfiguration() =
        member val TestName: string = null with get, set
        member val TestBed: string = null with get, set
        member val TestResultFolder: string = null with get, set
        member val OptionalFeaturedContainer: bool = true with get, set
        member val DotNet: DotNetSetting = new DotNetSetting() with get, set
        member val DiagTool: DiagToolSetting = new DiagToolSetting() with get, set
        member val TargetApp: TargetAppSetting = new TargetAppSetting() with get, set
        member val SystemInfo: SystemInformation = new SystemInformation() with get, set

    [<AbstractClass;Sealed>]
    type DiagToolsTestConfigurationGenerator private() =

        static member GenerateConfiguration(path: string) = 
            let _deserializer = (new DeserializerBuilder()).IgnoreUnmatchedProperties().Build()
            let serializedConfiguration = File.ReadAllText(path)

            try
                let configuration = _deserializer.Deserialize<DiagToolsTestConfiguration>(serializedConfiguration);
                let parseExn = new exn()
                if String.IsNullOrEmpty configuration.DotNet.SDKVersion
                then
                    parseExn.Data.Add(nameof(DiagToolsTestConfigurationGenerator), "Please specify .NET SDK version")
                    raise(parseExn)
                
                elif String.IsNullOrEmpty configuration.TestBed
                then
                    parseExn.Data.Add(nameof(DiagToolsTestConfigurationGenerator), "Please specify testbed")
                    raise(parseExn)
                
                elif String.IsNullOrEmpty configuration.DiagTool.DiagToolVersion
                then
                    parseExn.Data.Add(nameof(DiagToolsTestConfigurationGenerator), "Please specify diag tool version")
                    raise(parseExn)
                
                elif String.IsNullOrEmpty configuration.DiagTool.Feed
                then
                    parseExn.Data.Add(nameof(DiagToolsTestConfigurationGenerator), "Please specify diag tool feed")
                    raise(parseExn)

                elif String.IsNullOrEmpty configuration.SystemInfo.OSName
                then 
                    parseExn.Data.Add(nameof(DiagToolsTestConfigurationGenerator), "Please specify os name")
                    raise(parseExn)
                
                elif String.IsNullOrEmpty configuration.SystemInfo.CPUArchitecture
                then
                    parseExn.Data.Add(nameof(DiagToolsTestConfigurationGenerator), "Please specify processor architecture")
                    raise(parseExn)

                elif String.IsNullOrEmpty configuration.SystemInfo.CLIDebugger
                then
                    parseExn.Data.Add(nameof(DiagToolsTestConfigurationGenerator), "Please specify debugger")
                    raise(parseExn)

                else 
                    if String.IsNullOrEmpty(configuration.TestName)
                    then 
                        let testName = new StringBuilder("DiagTools")
                        testName.Append($"-{configuration.SystemInfo.OSName}") |> ignore
                        testName.Append($"-{configuration.SystemInfo.CPUArchitecture}") |> ignore
                        testName.Append($"-SDK{configuration.DotNet.SDKVersion}") |> ignore
                        testName.Append($"-Tool{configuration.DiagTool.DiagToolVersion}") |> ignore
                        let optionalFeaturedContainerFlag = 
                            match configuration.OptionalFeaturedContainer with
                            | true -> ""
                            | false -> "-NO"
                        testName.Append(optionalFeaturedContainerFlag) |> ignore
                        configuration.TestName <- testName.ToString()
                    configuration.TestResultFolder <- Path.Combine(configuration.TestBed, $"TestResult-{configuration.TestName}")

                    configuration.DotNet.DotNetRoot <- Path.Combine(configuration.TestBed, "dotnet-sdk")

                    configuration.DiagTool.ToolRoot <- Path.Combine(configuration.TestBed, "diag-tools")
                    
                    configuration.SystemInfo.EnvironmentVariables["DOTNET_ROOT"] <- configuration.DotNet.DotNetRoot
                    for de: DictionaryEntry in Environment.GetEnvironmentVariables() |> Seq.cast<DictionaryEntry> do
                        configuration.SystemInfo.EnvironmentVariables[de.Key.ToString()] <- de.Value.ToString()
                    if RuntimeInformation.RuntimeIdentifier.Contains("win")
                    then 
                        configuration.SystemInfo.EnvironmentVariables["Path"] <- 
                            Environment.GetEnvironmentVariable("Path") + $";{configuration.DotNet.DotNetRoot}"
                    else
                        configuration.SystemInfo.EnvironmentVariables["Path"] <- 
                            Environment.GetEnvironmentVariable("Path") + $":{configuration.DotNet.DotNetRoot}"

                    let targetAppsRoot = Path.Combine(configuration.TestBed, "TargetApps")
                    configuration.TargetApp.ConsoleApp <- new DotNetApp.DotNetApp(configuration.SystemInfo.EnvironmentVariables,
                                                                                  "console",
                                                                                  Path.Combine(targetAppsRoot, "console"))
                    configuration.TargetApp.WebApp <- new DotNetApp.DotNetApp(configuration.SystemInfo.EnvironmentVariables,
                                                                              "webapp",
                                                                              Path.Combine(targetAppsRoot, "webapp"))
                    configuration.TargetApp.GCDumpPlayground <- new DotNetApp.DotNetApp(configuration.SystemInfo.EnvironmentVariables,
                                                                                        "console",
                                                                                        Path.Combine(targetAppsRoot, "GCDumpPlayground"))
                

                configuration
            with _ -> 
                reraise()
         
            

File: src/DiagToolsValidationRunner/DiagToolsValidationRunner.Core/Configuration/DiagToolsTestRunConfiguration.cs
using System.Text;
using System.Collections;

using YamlDotNet.Serialization;

using DiagToolsValidationRunner.Core.Functionality;


namespace DiagToolsValidationRunner.Core.Configuration.DiagnosticsTest
{
    public class TestSetting: BaseTestSetting
    {
        public required bool OptionalFeatureContainer;
    }

    public class SystemInformation: BaseSystemInformation
    {
        public required string OSName;
        public required string CPUArchitecture;
        public required string CLIDebugger;
    }

    public class DiagToolSetting : DotNetToolSetting
    {
        public required List<string> ToolsToTest;
    }

    public class TargetAppSetting: BaseTargetAppSetting
    {
        public DotNetApp? ConsoleApp;
        public DotNetApp? WebApp;
        public DotNetApp? GCDumpPlayground;
    }

    public class DiagToolsTestRunConfiguration
    {
        public required TestSetting Test;
        public required DotNetSDKSetting SDKSetting;
        public required DiagToolSetting ToolSetting;
        public required TargetAppSetting AppSetting;
        public required SystemInformation SysInfo;
    }

    public static class DiagToolsTestConfigurationGenerator
    {
        private static IDeserializer _deserializer = (new DeserializerBuilder()).IgnoreUnmatchedProperties().Build();

        private static DiagToolsTestRunConfiguration ParseConfigFile(string configFile)
        {
            try
            {
                string serializedConfiguration = File.ReadAllText(configFile);
                DiagToolsTestRunConfiguration configuration =
                    _deserializer.Deserialize<DiagToolsTestRunConfiguration>(configFile);

                if (string.IsNullOrEmpty(configuration.SDKSetting.Version))
                {
                    throw new Exception(
                        $"{nameof(DiagToolsTestConfigurationGenerator)}: Please specify .NET SDK version");
                }
                else if (string.IsNullOrEmpty(configuration.Test.TestBed))
                {
                    throw new Exception(
                        $"{nameof(DiagToolsTestConfigurationGenerator)}: Please specify testbed");
                }

                else if (string.IsNullOrEmpty(configuration.ToolSetting.Version))
                {
                    throw new Exception(
                        $"{nameof(DiagToolsTestConfigurationGenerator)}: Please specify diag tool version");
                }

                else if (string.IsNullOrEmpty(configuration.ToolSetting.Feed))
                {
                    throw new Exception(
                        $"{nameof(DiagToolsTestConfigurationGenerator)}: Please specify diag tool feed");
                }

                else if (!(new List<string>{"Debug", "Release"}.Contains(configuration.AppSetting.BuildConfig)))
                {
                    throw new Exception(
                        $"{nameof(DiagToolsTestConfigurationGenerator)}: Please specify valid build config");
                }

                else if (string.IsNullOrEmpty(configuration.SysInfo.OSName))
                {
                    throw new Exception(
                        $"{nameof(DiagToolsTestConfigurationGenerator)}: Please specify OS name");
                }

                else if (string.IsNullOrEmpty(configuration.SysInfo.CPUArchitecture))
                {
                    throw new Exception(
                        $"{nameof(DiagToolsTestConfigurationGenerator)}: Please specify processor architecture");
                }

                else if (string.IsNullOrEmpty(configuration.SysInfo.CLIDebugger))
                {
                    throw new Exception(
                        $"{nameof(DiagToolsTestConfigurationGenerator)}: Please specify debugger");
                }

                return configuration;
            }
            catch
            {
                throw;
            }
        }

        public static DiagToolsTestRunConfiguration GenerateConfiguration(string configFile)
        {
            DiagToolsTestRunConfiguration configuration = ParseConfigFile(configFile);

            // Initialize Test
            StringBuilder testName = new();
            testName.Append($"-{configuration.SysInfo.OSName}");
            testName.Append($"-{configuration.SysInfo.CPUArchitecture}");
            testName.Append($"-SDK{configuration.SDKSetting.Version}");
            testName.Append($"-Tool{configuration.ToolSetting.Version}");
            if (configuration.Test.OptionalFeatureContainer)
            {
                testName.Append("");
            }
            else
            {
                testName.Append("-NO");
            }
            configuration.Test.TestName = testName.ToString();
            configuration.Test.TestResultFolder = Path.Combine(
                configuration.Test.TestBed, $"TestResult-{configuration.Test.TestName}");

            // Initialize SDKSetting
            configuration.SDKSetting.DotNetRoot = Path.Combine(
                configuration.Test.TestBed, "DotNetSDK");

            // Initialize ToolSetting
            configuration.ToolSetting.ToolRoot = Path.Combine(
                configuration.Test.TestBed, "DiagTools");

            // Initialize Environment
            configuration.SysInfo.EnvironmentVariables = new();
            foreach (DictionaryEntry de in Environment.GetEnvironmentVariables())
            {
                configuration.SysInfo.EnvironmentVariables[de!.Key!.ToString()!] = de!.Value!.ToString()!;
            }
            configuration.SysInfo.EnvironmentVariables["DOTNET_ROOT"] = configuration.SDKSetting.DotNetRoot;

            // Initialize Target Apps
            string targetAppsRoot = Path.Combine(configuration.Test.TestBed, "TargetApps");
            configuration.AppSetting.ConsoleApp = new(
                configuration.SysInfo.EnvironmentVariables, "console", Path.Combine(targetAppsRoot, "console"));
            configuration.AppSetting.WebApp = new(
                configuration.SysInfo.EnvironmentVariables, "webapp", Path.Combine(targetAppsRoot, "webapp"));
            configuration.AppSetting.GCDumpPlayground = new(
                configuration.SysInfo.EnvironmentVariables, "console", Path.Combine(targetAppsRoot, "GCDumpPlayground"));

            return configuration;
        }
    }
}
